modules/dnn/src/dnn.cpp

```diff
@@ -4976,6 +4976,13 @@ void Net::setPreferableTarget(int targetId)
             bool fp16 = ocl::Device::getDefault().isExtensionSupported("cl_khr_fp16");
             if (!fp16 && targetId == DNN_TARGET_OPENCL_FP16)
                 impl->preferableTarget = DNN_TARGET_OPENCL;
+#endif
+        }
+        else if (IS_DNN_CUDA_TARGET(targetId))
+        {
+#ifdef HAVE_CUDA
+            if (!cuda4dnn::doesDeviceSupportFP16())
+                impl->preferableTarget = DNN_TARGET_CUDA;
```

Comment:
@YashasSamaga Are you OK with proposed change?

Reply:
I don't think he is, my change patches the bug but is closer to a workeround rather than a proper fix

Reply:
The issue here is that this will involve the CUDA API before the diagnostic checks on the version, compatibility, etc. happen.

I think the bug, as @JulienMaille pointed out in another comment, is that the target for layers is set even before `initCUDABackend` is called. So changing the target inside `initCUDABackend` is useless.

One option would be to completely reinitialize the `net` object on fallback or redo the diagnostic checks in the `setPreferableTarget` before the FP16 check.